econia-labs/emojicoin-dot-fun
# [ECO-2618] Finalize emojicoin arena core logic design

# Description

This PR finalizes the Emojicoin Arena Move prototype, adding a README with a
comprehensive design overview.

Per extensive review, it adopts a lightweight data model that provides only the
minimal amount of information to required to enable comprehensive indexing,
as discussed in detail in the README.

## Logic updates

1. Add assorted inner abstractions to ease the review process.
1. Eliminate `SmartTable` structs since they can be a DoS vector.
1. Remove fields, events that are not required for lightweight data model.
1. Require that if a user has locked in since their most recent to an empty
   escrow, they must select lock in on subsequent top-off operations.
1. Allow user to elect lock-in even if they do not ultimately get matched, to
   prevent unexpected abortions.
1. Add exchange rate calculations.
1. Reset available rewards for a melee when it is cranked out.
1. Use integers instead of aggregators since exchange rate calculations require
   queries against both markets and thus inhibit parallelism within a melee
   in the general case.
1. Refactor type checking logic since exchange rate calculations do type checks.
1. Add assorted abstractions, helpers.
1. Add view functions and return unpackers for all core structs.
1. Add test-only fixturing to enable comprehensive state/event mocking.

## Test code

1. Add assorted mocking assertion functions.
1. Add additional test coin factories for `init_module` happy path tests.
1. Add `init_module` happy path tests for assorted pseudo-random seeds.
1. Add admin function happy path and expected failure tests

## Housekeeping

1. Use term `active` instead of `current` to indicate a melee that hasn't ended.
1. Update doc comments to link to actual structs, e.g. "`Escrow`" instead of
   "escrow".
1. Take out error codes used for admin-level guard rails that have since been
   removed.
1. Use compound assignment operators now that they are supported by `movefmt`
1. Change default melee time to 20 hours for new linear interpolation matching.

# Testing

From in `src/move/emojicoin_arena`:

```sh
git ls-files | entr -c sh -c " \
    aptos move compile --dev --move-2  &&
    aptos move fmt \
"
```

# Checklist
